Animation Docs (#759)
lambdaclass/curse_of_myrra: docs/src/animations.md

# Unity Animations

## Quick overview

#### The animator

The animator is the interface to control the [Animation System Overview](https://docs.unity3d.com/Manual/AnimationOverview.html). We use it to add the animations and create a flow within them.

#### Animation

The animation component is used to play animation clips, which are movement keyframes of the object we want to animate.

#### Transitions

Animation transitions allow the state machine to switch or blend from one animation state to another. We can use different parameters to make a transition happen. Such as booleans, floats, etc.

> Transitions have different settings to setup deppending on what you want, these are the most important.

Reviewer: ```suggestion
Transitions have different settings to setup depending on what you want, these are the most important.
```

- Has Exit Time: Determines a fixed time for an animation. Ignores any kind of transition parameter.
- If the Fixed Duration box is checked, the transition time is interpreted in seconds. If the Fixed Duration box is not checked, the transition time is interpreted as a fraction of the normalized time of the source state.
- Transition duration: This determines how much will it take to transition to the next state.

We need the Animator, animations and transitions to create solid animations with correct flow and states. Drag the animations inside the animator and create transition (right click in the state) between them to start using your animations.

## How we handle animations

### Parameters

We use a specific list of parameters for our transitions

> - `SkillBasic, Skill1, Skill2, Skill3, Skill4 ( Booleans )` We use them to change the state of the animation and make the transition happen.

Reviewer: ```suggestion
- `SkillBasic, Skill1, Skill2, Skill3, Skill4 (Booleans)` We use them to change the state of the animation and make the transition happen.
```

- `SkillBasicSpeed_start, Skill1Speed_start, Skill2Speed_start, Skill3Speed_start, Skill4Speed_start (Floats)` We use them to control the speed of the animation start. To use them you have to set the animation to use the multiplier parameter and choose the respective parameter.

![](./videos/parameterCustom.gif)

### Scripts

Let's dive into how we use the animations to match with our backend, the scripts we use and more. The main scripts that participate in the animation flow are:

- `Skillinfo.cs` a ScriptableObject with the skill information, here what we care about are the fields `hasModelAnimation`, `startAnimationDuration`, `executeAnimationDuration`, and `animationSpeedMultiplier` fields. We'll talk about them later, so keep them in mind.

- `Skill.cs` This is a really important script, here we control the complete flow of the animations (Play, Stop, Block movements, etc)

- `SkillAnimationEvents.cs` is in charge of changing the active skill playing and end the animation playing.
- `Battle.cs` Depending on the backend action, executes the corresponding skill and starts the flow of the animation system.

> - `CustomLevelManager.cs` Where the buttons and mapped with the respective skill.

Reviewer 2: ```suggestion
- `CustomLevelManager.cs` Where the buttons are mapped with the respective skill.
```

# Unity Animations

## Quick overview

#### The animator

> The animator is the interface to control the Mecanim animation system. We use it to add the animations and create a flow within them.

Reviewer 3: ```suggestion
The animator is the interface to control the [Mecanim animation system](https://docs.unity3d.com/462/Documentation/Manual/MecanimAnimationSystem.html). We use it to add the animations and create a flow within them.
```

#### Animation

The animation component is used to play animation clips, which are movement keyframes of the object we want to animate.

#### Transitions

Animation transitions allow the state machine to switch or blend from one animation state to another. We can use differents parameters to make a transition happend. Such as booleans, floats, etc.

> Transitions have diffenrents settings to setup deppending of what you want, these are the most important.

Reviewer 3: ```suggestion
Transitions have differents settings to setup deppending on what you want, these are the most important.
```

> - Has Exit Time: Determinates a fixed time for a animation. Ignores any kind of transition parameter.

Reviewer 2: *an animation

- If the Fixed Duration box is checked, the transition time is interpreted in seconds. If the Fixed Duration box is not checked, the transition time is interpreted as a fraction of the normalized time of the source state.
- Transition duration: This determinates how much will it took to transition to the next state.

We need The Animator, the animations and transitions to create solid animations with correct flow and states. Drag the animations inside the animator and create transition (right click in the state) between them to start using your animations.

## How we handle animations

### Parameters

We use a specific list of parameters for our transitions

- `SkillBasic, Skill1, Skill2, Skill3, Skill4 ( Booleans )` We use them to change the state of the animation and make the transition happend.

> - `SkillBasic_start, Skill1c_start, Skill2c_start, Skill3c_start, Skill4c_start ( Booleans )` Not allways necessaries , we use them to concatenate animations.

Reviewer 2: *Not always necessary

Reviewer 2: Not sure if I follow what these booleans do by just reading this. (SkillBasic_start, Skill1c_start, Skill2c_start, Skill3c_start, Skill4c_start)

- `SkillBasicSpeed, Skill1Speed, Skill2Speed, Skill3Speed, Skill4Speed ( Floats )` We use them to control the speed of the animation. To use them you have to set the animation to use multiplier parameter and choose the respective parameter.

- `SkillBasicSpeed_start, Skill1Speed_start, Skill2Speed_start, Skill3Speed_start, Skill4Speed_start ( Floats )` We use them to control the speed of the animation start. To use them you have to set the animation to use multiplier parameter and choose the respective parameter.

![](./videos/parameterCustom.gif)

### Scripts

Lets deep into how we use the animations to match with our backend, the scripts we use and more. The main scripts that participate in the animation flow are:

> - `Skillinfo.cs` an ScriptableObject with the skill information, here what we care about are the fields `hasModelAnimation`, `startAnimationDuration`, `executeAnimationDuration` and `animationSpeedMultiplier` fields. Lets talk about them later, keep them in mind.

Reviewer 2: *a ScriplableObject. If the next word starts with the sound of a consonant then it should be a, if it starts with the sound of a vowel its an

- `Skill.cs` This is a really importart script, here we control the complete flow of the animations ( Play, Stop, Block movements, etc )

> - `SkillAnimationEvents.cs` is in change of changing the active skill playing and end the animation playing.

Reviewer 2: *it is in charge

- `Battle.cs` Depending of the backend action executes the corresponding skill and start the flow of the animation system.

> - `CustomLevelManager.cs` Where the buttons and mapped with the respective skill.

Reviewer 2: Maybe I would include the route to this scripts not just the name of the script

Author: By route do you mean for example `Assets/Scripts/Skills/Skill.cs` ?

### SkillInfo fields

- `hasModelAnimation`: (``boolean``) Allows the animation to start and execute. If this is not true you would have to start the animation manually. It can be useful to have it false with animations like walking where you want them to play constantly.

- `executeAnimationDuration`: (``float``) Determinates the time that the animation will be played.

- `startAnimationDuration`: (``float``) Determinates the time to start the animation.

> - `animationSpeedMultiplier`: (``float``) Determinate the play speed of the animation (used in the transition parameters if it is set).

Reviewer 2: *Determinates

### Skill.cs

> In this scripts is where all the skill-animation flows happen. It has all the scripts to start, execute, stop and change the animations. I believe this script leads to a separeted guide. Check the documentation [Skill docs](./skill.md).

Reviewer 2: *script, happens

- `SkillBasic, Skill1, Skill2, Skill3, Skill4 ( Booleans )` We use them to change the state of the animation and make the transition happend.

> - `SkillBasic_start, Skill1c_start, Skill2c_start, Skill3c_start, Skill4c_start ( Booleans )` Not always necessary , we use them to concatenate animations. For example if a skill have 2 different animations that happends sequentially and we want to concatenate them to create the whole animation we can do something like this:

Reviewer 2: For example, if a skill has two different animations that happen ...

### SkillInfo fields

> Tiempo maximo de duracion para hacer un eject para ambas

Reviewer 2: This is in spanish

- `executeAnimationDuration`: (``float``) Determinates the time that the animation will be played.

- `startAnimationDuration`: (``float``) Determinates the time to start the animation.

> - `startAnimationDuration`: (``float``) and `executeAnimationDuration`: (``float``) These methods represent the maxinum duration time to eject the skill animation

Reviewer 2: maximum
